flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtils.java

```diff
@@ -0,0 +1,659 @@
+
+	// ------------------------------------------------------------------------
+	//  Generating JVM Parameters
+	// ------------------------------------------------------------------------
+
+	public static String generateJvmParametersStr(final TaskExecutorResourceSpec taskExecutorResourceSpec) {
+		final MemorySize jvmHeapSize = taskExecutorResourceSpec.getFrameworkHeapSize()
+			.add(taskExecutorResourceSpec.getTaskHeapSize())
+			.add(taskExecutorResourceSpec.getOnHeapManagedMemorySize());
+		final MemorySize jvmDirectSize = taskExecutorResourceSpec.getTaskOffHeapSize()
+			.add(taskExecutorResourceSpec.getShuffleMemSize());
+		final MemorySize jvmMetaspaceSize = taskExecutorResourceSpec.getJvmMetaspaceSize();
+
+		return "-Xmx" + jvmHeapSize.getBytes()
+			+ " -Xms" + jvmHeapSize.getBytes()
+			+ " -XX:MaxDirectMemorySize=" + jvmDirectSize.getBytes()
+			+ " -XX:MetaspaceSize=" + jvmMetaspaceSize.getBytes();
+	}
+
+	// ------------------------------------------------------------------------
+	//  Generating Dynamic Config Options
+	// ------------------------------------------------------------------------
+
+	public static String generateDynamicConfigsStr(final TaskExecutorResourceSpec taskExecutorResourceSpec) {
+		final Map<String, String> configs = new HashMap<>();
+		configs.put(TaskManagerOptions.FRAMEWORK_HEAP_MEMORY.key(), taskExecutorResourceSpec.getFrameworkHeapSize().getBytes() + "b");
+		configs.put(TaskManagerOptions.TASK_HEAP_MEMORY.key(), taskExecutorResourceSpec.getTaskHeapSize().getBytes() + "b");
+		configs.put(TaskManagerOptions.TASK_OFF_HEAP_MEMORY.key(), taskExecutorResourceSpec.getTaskOffHeapSize().getBytes() + "b");
+		configs.put(TaskManagerOptions.SHUFFLE_MEMORY_MIN.key(), taskExecutorResourceSpec.getShuffleMemSize().getBytes() + "b");
+		configs.put(TaskManagerOptions.SHUFFLE_MEMORY_MAX.key(), taskExecutorResourceSpec.getShuffleMemSize().getBytes() + "b");
+		configs.put(TaskManagerOptions.MANAGED_MEMORY_SIZE.key(), taskExecutorResourceSpec.getManagedMemorySize().getBytes() + "b");
+		configs.put(TaskManagerOptions.MANAGED_MEMORY_OFFHEAP_SIZE.key(), taskExecutorResourceSpec.getOffHeapManagedMemorySize().getBytes() + "b");
+		return assembleDynamicConfigsStr(configs);
+	}
+
+	private static String assembleDynamicConfigsStr(final Map<String, String> configs) {
+		final StringBuilder sb = new StringBuilder();
+		for (Map.Entry<String, String> entry : configs.entrySet()) {
+			sb.append("-D ").append(entry.getKey()).append("=").append(entry.getValue()).append(" ");
+		}
+		return sb.toString();
+	}
+
+	// ------------------------------------------------------------------------
+	//  Memory Configuration Calculations
+	// ------------------------------------------------------------------------
+
+	public static TaskExecutorResourceSpec resourceSpecFromConfig(final Configuration config) {
+		if (isTaskHeapMemorySizeExplicitlyConfigured(config) && isManagedMemorySizeExplicitlyConfigured(config)) {
```

Comment:
Why does the managed memory needs to be explicitly configured? Can't we calculate it via `MANAGED_MEMORY_FRACTION` if it's not set?

Reply:
We can derive managed memory size from its fraction in either of the following two ways:
- `managed_size = total_flink_size * managed_fraction`
- `managed_size = remaining_flink_internal_size / (1 - managed_fraction) * managed_fraction`

If total flink memory is set or can be derived from total process memory, we use the first equation to derive managed memory size in later 'else-if' branches.

For this 'if' branch, if managed memory size is not explicitly configured and we want to derive it from its fraction, we will have to use the second equation. In that case, we need `remaining_flink_internal_size` to be determinate, which may not always be the case given that shuffle memory size can also not be explicitly configured.

There are some alternatives that I can think of, but I'm not sure whether it's necessary to go with either of them, given the complexity for user to understand and for the maintainability.
- We can derive managed and shuffle memory size together, with the following equations
  - `managed_and_shuffle_size = remaining_flink_internal_size / (1 - managed_fraction - shuffle_fraction) * (managed_fraction + shuffle_fraction)`
  - `managed_size = managed_and_shuffle_size / (managed_fraction + shuffle_fraction) * managed_fraction`
  - `shuffle_size = managed_and_shuffle_size / (managed_fraction + shuffle_fraction) * shuffle_fraction`
But given that shuffle memory size also need to respect to the min/max, this approach can be even more complicated.
- We can require either managed memory size or shuffle memory size to be explicitly configured, and derive the other with its fraction. This also complicates the calculation logic by introducing more 'if-else' branches.

Reply:
Ok, then we say as a limitation one needs to also configure the managed memory if one has the task heap memory configured. I guess this should then be properly documented.

Reply:
Sure.